contracts/riskon/UniswapV3RiskOnVault.sol

```diff
@@ -0,0 +1,673 @@
+                    //add liquidity
+                    nonfungiblePositionManager.increaseLiquidity(INonfungiblePositionManager.IncreaseLiquidityParams({
+                    tokenId : limitMintInfo.tokenId,
+                    amount0Desired : _balance0,
+                    amount1Desired : _balance1,
+                    amount0Min : 0,
+                    amount1Min : 0,
+                    deadline : block.timestamp
+                    }));
+                }
+            }
+        }
+        netMarketMakingAmount += _amount;
+        emit LendToStrategy(_amount);
+    }
+
+    /// @notice Redeem
+    /// @param _redeemShares The amount of shares to withdraw
+    /// @param _totalShares The total amount of shares owned by this strategy
+    /// @return _redeemBalance The balance of redeem
+    function redeem(uint256 _redeemShares, uint256 _totalShares) external isOwner override returns (uint256 _redeemBalance) {
+        _redeemBalance = redeemToVault(_redeemShares, _totalShares);
+        if (_redeemBalance > netMarketMakingAmount) {
+            netMarketMakingAmount = 0;
+        } else {
+            netMarketMakingAmount -= _redeemBalance;
+        }
+        IERC20Upgradeable(wantToken).safeTransfer(msg.sender, _redeemBalance);
+        emit Redeem(_redeemBalance);
+    }
+
+    /// @notice Redeem to vault by keeper
+    /// @param _redeemShares The amount of shares to withdraw
+    /// @param _totalShares The total amount of shares owned by this strategy
+    /// @return _redeemBalance The balance of redeem
+    function redeemToVaultByKeeper(uint256 _redeemShares, uint256 _totalShares) external override returns (uint256 _redeemBalance) {
+        _redeemBalance = redeemToVault(_redeemShares, _totalShares);
+        emit RedeemToVault(_redeemBalance);
+    }
+
+    /// @notice Redeem to vault
+    /// @param _redeemShares The amount of shares to withdraw
+    /// @param _totalShares The total amount of shares owned by this strategy
+    /// @return _redeemBalance The balance of redeem
+    function redeemToVault(uint256 _redeemShares, uint256 _totalShares) internal whenNotEmergency nonReentrant returns (uint256 _redeemBalance) {
+        uint256 currentBorrow = uniswapV3RiskOnHelper.getCurrentBorrow(borrowToken, interestRateMode, address(this));
+        (uint256 _totalCollateral, , , , ,) = uniswapV3RiskOnHelper.borrowInfo(address(this));
+        address aCollateralToken = uniswapV3RiskOnHelper.getAToken(wantToken);
+        if (_redeemShares == _totalShares) {
+            harvest();
+            withdraw(baseMintInfo.tokenId, _redeemShares, _totalShares);
+            withdraw(limitMintInfo.tokenId, _redeemShares, _totalShares);
+            delete baseMintInfo;
+            delete limitMintInfo;
+            uint256 borrowTokenBalance = balanceOfToken(borrowToken);
+            if (currentBorrow > borrowTokenBalance) {
+                IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, 0);
```

Comment:
can put L318 and 319 into function initialize,approve type(uint256).max amount for gas save
```
 IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, type(uint256).max);
```